README.md
# sysmon
A lightweight system monitoring solution

## Features
- implemented in bash + sqlite3
- KPI based customzation
- support local host, remote host with ssh key and remote host with user/pass
- alarm db to track history alarms 
- support cli: sysmon, localmon, remotemon, hostping

## test env
```
wsl ubuntu on win10
127.0.0.1 localhost wsl1 wsl2 
```
add wslx as remote server just for test, it enabled user/password login and ssh key

## Usages
```
/shared/sysmon$ ./sysmon.sh
Usage: ./sysmon.sh [-s localhost|server] [-f serverfile] [-k all|kpi] [-m key|pass] [-p] [-d] [-e emailist] [-l] [-h]
-s server string, localhost need -k option; remote server need -k -m options for system monitor or needs -p for ping test
-f server filename, cannot exist with -s at sametime; need -k -m options for system monitor or need -p for ping test
-k kpi name, all or valid kpi name like fs, mem and cpu etc
-m access mode, ssh remote access with key or user/password
-p ping flag, need to work with -s or -f option
-d debug flag
-e email list, will use default list if leave as ''
-l list available kpi list
-h help
```

## check available kpi 
```
/shared/sysmon$ ./sysmon.sh -l
fs mem cpu
```
## host ping scan 
```
/shared/sysmon$./sysmon.sh -p -s localhost
03-02-2020-13-21-48   localhost  ping - [ Server  localhost  : up  ]

/shared/sysmon$ ./sysmon.sh -p -s "wsl1 wsl2"
03-02-2020-13-22-51   wsl1  ping - [ Server  wsl1  : up  ]
03-02-2020-13-22-51   wsl2  ping - [ Server  wsl2  : up  ]

/shared/sysmon$ ./sysmon.sh -p -f nodelist
03-02-2020-13-23-03   wsl1  ping - [ Server  wsl1  : up  ]
03-02-2020-13-23-03   wsl2  ping - [ Server  wsl2  : up  ]

/shared/sysmon$ ./sysmon.sh -p -s wsl3  
03-02-2020-13-29-55   wsl3   ping   HIGH [  Server wsl3 : down  ]
```
## monitor for localhost, all KPI
```
/shared/sysmon$ ./sysmon.sh -s localhost -k all
02-29-2020-18-18-05   localhost   fs   HIGH [  rootfs 511G 444G 67G 87% /  ]
02-29-2020-18-18-05   localhost   fs   HIGH [  none 511G 444G 67G 87% /dev  ]
02-29-2020-18-18-05   localhost   fs   HIGH [  none 511G 444G 67G 87% /run  ]
02-29-2020-18-18-05   localhost   fs   HIGH [  none 511G 444G 67G 87% /run/lock  ]
02-29-2020-18-18-05   localhost   fs   HIGH [  none 511G 444G 67G 87% /run/shm  ]
02-29-2020-18-18-05   localhost   fs   HIGH [  none 511G 444G 67G 87% /run/user  ]
02-29-2020-18-18-05   localhost   fs   HIGH [  C: 511G 444G 67G 87% /mnt/c  ]
02-29-2020-18-18-05   localhost   mem   HIGH [  memory total: 32GB memory used: 13GB memory free: 18GB memory usage: 42%  ]
02-29-2020-18-18-05   localhost   cpu   HIGH [  cpu usage(>5%): 8%  ]
```

## monitor for remote server with key on selected kpi
```
/shared/sysmon$ ./sysmon.sh -s wsl -k "mem cpu" -m key
02-29-2020-18-21-57   wsl   mem   HIGH [  memory total: 32GB memory used: 13GB memory free: 18GB memory usage: 42%  ]
02-29-2020-18-21-57   wsl   cpu   HIGH [  cpu usage(>5%): 8%  ]
```

## monitor for remote server with user/pass on selected kpi 
```
/shared/sysmon$ ./sysmon.sh -s wsl -k "cpu fs" -m pass
02-29-2020-18-24-03   wsl   cpu   HIGH [  cpu usage(>5%): 8%  ]
02-29-2020-18-24-03   wsl   fs   HIGH [  rootfs 511G 444G 67G 87% /  ]
02-29-2020-18-24-03   wsl   fs   HIGH [  none 511G 444G 67G 87% /dev  ]
02-29-2020-18-24-03   wsl   fs   HIGH [  none 511G 444G 67G 87% /run  ]
02-29-2020-18-24-03   wsl   fs   HIGH [  none 511G 444G 67G 87% /run/lock  ]
02-29-2020-18-24-03   wsl   fs   HIGH [  none 511G 444G 67G 87% /run/shm  ]
02-29-2020-18-24-03   wsl   fs   HIGH [  none 511G 444G 67G 87% /run/user  ]
02-29-2020-18-24-03   wsl   fs   HIGH [  C: 511G 444G 67G 87% /mnt/c  ]
```
## monitor for remote server with ssh key for server list file
```
/shared/sysmon$ ./sysmon.sh -f config/nodelist -k cpu -m key
03-02-2020-13-24-58   wsl1   cpu   HIGH [  cpu usage(>5%): 8%  ]
03-02-2020-13-24-58   wsl2   cpu   HIGH [  cpu usage(>5%): 8%  ]
```
## monitor for remote server with user/pass for server list file
```
/shared/sysmon$ ./sysmon.sh -f config/nodelist -k mem -m pass
03-02-2020-13-25-20   wsl1   mem   HIGH [  memory total: 32GB memory used: 13GB memory free: 19GB memory usage: 41%  ]
03-02-2020-13-25-20   wsl2   mem   HIGH [  memory total: 32GB memory used: 13GB memory free: 19GB memory usage: 41%  ]
```
## hostping cli 
```
/shared/sysmon$ ./hostping.sh -p -s wsl2
03-03-2020-18-38-04 wsl2 ping - [ Server wsl2 : up ]

/shared/sysmon$ ./hostping.sh -p -f config/nodelist
03-03-2020-18-38-13 wsl1 ping - [ Server wsl1 : up ]
03-03-2020-18-38-13 wsl2 ping - [ Server wsl2 : up ]
```
## remotemon cli 
```
/shared/sysmon$ ./remotemon.sh -k cpu -m key -s wsl1
03-03-2020-18-42-27   wsl1   cpu   HIGH [  cpu usage(>5%): 9%  ]

/shared/sysmon$ ./remotemon.sh -k cpu -m key -f config/nodelist
03-03-2020-18-42-36   wsl1   cpu   HIGH [  cpu usage(>5%): 9%  ]
03-03-2020-18-42-36   wsl2   cpu   HIGH [  cpu usage(>5%): 9%  ]
```
## localmon cli 
```
/shared/sysmon$ ./localmon.sh -k all
03-03-2020-19-30-39   localhost   fs   HIGH [  rootfs 511G 446G 65G 88% /  ]
03-03-2020-19-30-39   localhost   fs   HIGH [  none 511G 446G 65G 88% /dev  ]
03-03-2020-19-30-39   localhost   fs   HIGH [  none 511G 446G 65G 88% /run  ]
03-03-2020-19-30-39   localhost   fs   HIGH [  none 511G 446G 65G 88% /run/lock  ]
03-03-2020-19-30-39   localhost   fs   HIGH [  none 511G 446G 65G 88% /run/shm  ]
03-03-2020-19-30-39   localhost   fs   HIGH [  none 511G 446G 65G 88% /run/user  ]
03-03-2020-19-30-39   localhost   fs   HIGH [  C: 511G 446G 65G 88% /mnt/c  ]
Tue Mar  3 19:30:40 STD 2020 - localhost - MEM
03-03-2020-19-30-39   localhost   mem   HIGH [  memory total: 32GB memory used: 13GB memory free: 18GB memory usage: 42%  ]
03-03-2020-19-30-39   localhost   cpu   HIGH [  cpu usage(>5%): 9%  ]
```

## turn on debug mode
```
/shared/sysmon$ ./sysmon.sh -d -k mem -s localhost
03-04-2020-14-38-45   localhost   mem   HIGH [  memory total: 32GB memory used: 14GB memory free: 18GB memory usage: 42%  ]
Please see detail log at /shared/sysmon/outlog/<kpi>_<node>_03-04-2020-14-38-45.log
The system monitoring report saved at /shared/sysmon/report/sysmon_report_03-04-2020-14-38-45.log
```

## enable email 
```
/shared/sysmon$ ./sysmon.sh -d -e "info@sysmon.com" -k mem -m key -s wsl2
03-04-2020-14-38-19   wsl2   mem   HIGH [  memory total: 32GB memory used: 14GB memory free: 18GB memory usage: 43%  ]
Please see detail log at /shared/sysmon/outlog/<kpi>_<node>_03-04-2020-14-38-19.log
The system monitoring report saved at /shared/sysmon/report/sysmon_report_03-04-2020-14-38-19.log
send_email System monitoring report - 03-04-2020-14-38-19 /shared/sysmon/report/sysmon_report_03-04-2020-14-38-19.log HIGH info@sysmon.com

/shared/sysmon$ ./localmon.sh -d -e '' -k cpu
03-04-2020-16-08-44   localhost   cpu   HIGH [  cpu usage(>5%): 7%  ]
Please see detail log at /shared/sysmon/outlog/<kpi>_<node>_03-04-2020-16-08-44.log
The system monitoring report saved at /shared/sysmon/report/sysmon_report_03-04-2020-16-08-44.log
send_email System monitoring report - 03-04-2020-16-08-44 /shared/sysmon/report/sysmon_report_03-04-2020-16-08-44.log HIGH demo@sysmon.com
```

changelog.md
# changelog for sysmon 

## Feb 20, 2020  
v1.1 kpi for FS space usage

## Feb 22, 2020  
v1.2 supports kpi loop
v1.3 add alarm, email functions

## Feb 23, 2020 
v1.4 add sqlite3 alarmdb

## Feb 29, 2020
v1.5 add getopts for cli

## March 1st, 2020
v1.6
add host ping -p and -f filename
move kpi customization to sysmon-kpi.bash 

## Mar 3rd,2020
v1.7 move common function to sysmon-common.sh,sysmon-kpi.sh,alarm-common.sh under ./util  

## Mar 4th, 2020 
- rename/rewrite validkpi() to general function belongset(), move to sysmon-common
- add -d debug flag 
- add -e emaillist 





sysmon.sh
#!/bin/bash
# system monitoring solution
# https://github.com/robertluwang
# v1.1 kpi for FS space usage  Feb 20, 2020 
# v1.2 supports kpi loop  Feb 22, 2020 
# v1.3 add alarm, email functions  Feb 22, 2020 
# v1.4 add sqlite3 alarmdb Feb 23, 2020 
# v1.5 add getopts for cli  Feb 29, 2020
# v1.6 March 1st, 2020
# add host ping -p and -f filename
# move kpi customization to sysmon-kpi.bash 
# v1.7 move common function to sysmon-common.sh,sysmon-kpi.sh,alarm-common.sh under ./util  Mar 3rd,2020

## import common functions 

SYSMON=`pwd`
source $SYSMON/util/sysmon-common.sh
source $SYSMON/util/sysmon-kpi.sh
source $SYSMON/util/alarm-common.sh

# disable debug and email 
DEBUG=0
EMAILYES=0

## main loop 

unset SERVER FILE KPI MODE 

PING=0

usage()
{
    echo "Usage: $0 [-s localhost|server] [-f serverfile] [-k all|kpi] [-m key|pass] [-p] [-d] [-e emaillist] [-l] [-h]"
    echo "-s server string, localhost need -k option; remote server need -k -m options for system monitor or need -p for ping test"
    echo "-f server filename, cannot exist with -s at sametime; need -k -m options for system monitor or need -p for ping test"
    echo "-k kpi name, all or valid kpi name like fs, mem and cpu etc"
    echo "-m access mode, ssh remote access with key or user/password"
    echo "-p ping flag, need to work with -s or -f option"
    echo "-d debug flag"
    echo "-e email list, will use default list if leave as ''"
    echo "-l list available kpi list"
    echo "-h help"
    exit
}

if [ $# -eq 0 ]; then
    usage
fi

while getopts ":s:f:k:m:e:pdlh" opt; do
case $opt in
    s) SERVER="$OPTARG"
    if [[ -z "$SERVER" ]];then 
        echo "Cannot be empty for server"
        usage
    fi
    ;;
    f) FILE="$OPTARG"
    if [[ -z "$FILE" ]];then 
        echo "Cannot be empty for server file"
        usage
    fi
    ;;
    k) KPI="`upcase "$OPTARG"`"
    if [[ "$KPI" == "ALL" ]];then 
        KPI="$KPIFULL"  
    elif [[ $(belongset "$KPI" "$KPIFULL") == 0 ]];then
        echo Invalid kpi $KPI
        usage
    fi 
    ;;
    m) MODE="`upcase "$OPTARG"`"
    if [[ "$MODE" == "PASS" ]] || [[ "$MODE" == "KEY" ]];then
        :
    else
        echo Invalid access mode $MODE
        usage
    fi
    ;;
    p) PING=1
    ;;
    d) DEBUG=1
    ;;
    e) [[ -n "$OPTARG"  ]] && EMAIL_LIST="$OPTARG"
    EMAILYES=1
    ;;
    l) echo `lowcase "$KPIFULL"`
    exit
    ;;
    h) usage
    ;;
    *) echo Invalid options
    usage
    ;; 
esac
done

## rules for options check 

if [[ $PING -eq 1 ]];then 
    if [[ -n $KPI || -n $MODE ]];then
        echo "1 No -k and -m needed for ping test"
        usage
    elif [[ -z $FILE ]] && [[ -z $SERVER ]];then
        echo "2 -s|-f needed for ping test"
        usage
    elif [[ -n $FILE ]] && [[ -n $SERVER ]];then
        echo "3 Cannot have -s and -f at sametime"
        usage
    fi
fi

if [[ $PING -eq 0 ]] && [[ -n $KPI ]] && [[ -n $MODE ]];then
    if [[ -z $SERVER ]] && [[ -z $FILE ]];then
        echo "4 -s|-f needed for system monitor"
        usage
    fi
fi

if [[ $PING -eq 0 && -n $KPI && -z $MODE && ( -n $SERVER || -n $FILE ) ]];then
    if [[ "$SERVER" == "localhost" ]];then
        :
    else
        echo "5 -m needed for remote access"
        usage
    fi
fi

if [[ $PING -eq 0 ]] && [[ "$SERVER" == "localhost" ]];then
    if [[ -n $MODE ]];then
        echo "6 no -m needed for localhost monitor"
        usage
    elif [[ -n $FILE ]];then 
        echo "7 Cannot have -f with -s at sametime"
        usage
    elif [[ -z $KPI ]];then 
        echo "8 -k needed for localhost monitor"
        usage
    fi
fi

if [[ $PING -eq 0 && -z $KPI && -n $MODE && ( -n $SERVER || -n $FILE ) ]];then
    echo "9 -k needed for system monitor"
    usage
fi

if [[ $PING -eq 0 && -z $KPI && -z $MODE && ( -n $SERVER || -n $FILE ) ]];then
    echo "10 -k and -m needed for system monitor"
    usage
fi

if [[ $PING -eq 0 && -n $KPI ]];then
    if [[ "$SERVER" == "localhost" ]];then
        :
    elif [[ -z $MODE || ( -z $SERVER && -z $FILE ) ]];then
        echo "11 -m or -s|-f needed for system monitor"
        usage
    fi
fi

if [[ $PING -eq 0 && -n $MODE ]];then
    if [[ -z $KPI || ( -z $SERVER && -z $FILE ) ]];then
        echo "12 -k or -s|-f needed for system monitor"
        usage
    fi
fi


## host ping scan task
if [[ $PING -eq 1 ]];then 
    if [[ -n $FILE ]];then
        host_ping "$FILE"
        exit
    else
        host_ping "$SERVER"
        exit
    fi
fi

## system monitor task 

if [[ $PING -eq 0 && -n $SERVER && -z $FILE ]];then
    if [[ -n $KPI && -z $MODE ]];then
        if [[ "$SERVER" == "localhost" ]];then
            local_monitor
        fi 
    fi
fi

if [[ $PING -eq 0 && -n $KPI && -n $MODE && ( -n $SERVER || -n $FILE ) ]];then
    if [[ $MODE == "KEY" ]]; then
        if [[ -n $FILE ]];then
            remote_monitor_key "$FILE"
        else
            remote_monitor_key "$SERVER"
        fi
    elif [[ $MODE == "PASS" ]]; then 
        if [[ -n $FILE ]];then
            remote_monitor_pass "$FILE"
        else
            remote_monitor_pass "$SERVER"
        fi
    fi
fi



    